### PR TITLE
Add support for tool length offsets (G10 L1/G10 L10/G43/G49)

### PR DESCRIPTION
**Pull request by lutorm**
It would be painful to not have tool length offsets for my application, so I attempted to add them. It appears to work correctly, but has not seen extensive use yet. If there are things that could be done better, please suggest improvements.

### CODE REVIEW

**Comment by justinclift** [comment]
@lutorm Just noticed the 2nd commit, fixing a path change from the first commit.

As a thought, you can easily change your original commit instead to having to do a 2nd one as follow up. :smile:

To do that, use a "git reset xxxxxx" command (**without** the --hard option) to move the index just before your initial commit.  Then the "git add <files>" then "git commit" again, which will make a new commit with just the changes you want.  Then "git push <repo/branch bits> --force" to push the new commit here to this existing PR.  GitHub's code is smart enough to understand what you're trying to do, and will update this PR to just have the new commit you pushed.

Does that make sense? :smile:

**Comment by lutorm** [comment]
@justinclift Ok, cleaned up the commit. FWIW, I've been running this patch since Dec and it seems to work correctly at least in my use case.

**Comment by justinclift** [comment]
@lutorm Cool.  It seems like PR's don't get merged here though. Synthetos seem to be using Open Source licensing, but missing the entire "Community Building" activity around that. 😦

**Comment by aldenhart** [comment]
Our bad. We have been so head down on some major work that we have not been good about interacting. We are preparing a mjor edge push and will go through all the PRs to bring in what we can. We do appreciate the work, and need to be better about this.

**Comment by aldenhart** [comment]
@lutorm do you mind if we manually move this in (cherry pick) to the new edge push or would you prefer to wait for that push and do another PR against it?

**Comment by lutorm** [comment]
@aldenhart Please feel free to make use of it in whatever way feels best to you.

I had one question, though: I made the size of the tool table 8 entries. This isn't enough to support even my small hobbyist setup, so unless there are memory issues (which I think there aren't on the G2 hardware) it would be preferable to increase this to something much larger, like 128 entries. At that point it sort of becomes infeasible to enumerate them in the settings as is done currently, and you really need a way to persistently save the settings. Thoughts?

File: TinyG2/canonical_machine.cpp [hunk 1]
```diff
@@ -893,35 +893,129 @@ stat_t cm_set_coord_offsets(const uint8_t coord_system,
...
+            // you can't set G53
+            return (STAT_P_WORD_IS_INVALID);
+        }
+
+        for (uint8_t axis = AXIS_X; axis < AXES; axis++) {
+            if (flag[axis]) {
+                if (L_word == 2) {
+                    cm.offset[coord_system][axis] = _to_millimeters(offset[axis]);
+                } else {
+                    // Should L20 take into account G92 offsets?
```
Comment (lutorm) [inline TinyG2/canonical_machine.cpp]: Does anyone know what the correct G-code behavior is here?

**Comment by aldenhart** [comment]
We are working on a more general concept for tool tables and other persistence via the JSON interface. You would load and save them as files on the host. But that's a ways out. how many do you need in the mean time? somewhere between 8 and 128

**Comment by lutorm** [comment]
How about we make it 32?

**Comment by aldenhart** [comment]
@lutorm - Can you please give us a way to contact you directly? There is no email in your github profile. You can use the Synthetos gmail address or customerservice at synthetos dot com address to contact us. I have some details of the tool table I wanted to ask you. I will publish the results back on this forum. Thanks.

**Comment by justinclift** [comment]
@aldenhart, @lutorm's email address is in the commit data.  eg using "git log" from the command line shows it.

**Comment by aldenhart** [comment]
Thanks

**Comment by lutorm** [comment]
Sorry, didn't realize. I've added the email to the profile.

**Comment by lutorm** [comment]
@aldenhart Did you email me about this? I haven't gotten anything.

**Comment by aldenhart** [comment]
@lutorm , I'm working on getting all this out for test. I have your code installed and have been running it. I will post as a dev branch, probably something like dev-109-tool-offset - hopefully today if I do not get side tracked. 

One question - I set the tool table size to 16. Is this adequate for your uses, or do you need it to be larger? I can expand to 24 or 32, but bigger than this is problematic.

**Comment by lutorm** [comment]
Cool, looking forward to trying it out. I saw you had added a wiki page for
it. (And noted that my code did not handle H0 like Tormach...)

I've been working with John Lauer to improve the handling of tool changes
in chilipeppr, too. I'm getting close to something I'm quite satisfied with.

It's funny you should ask, I expanded my branch to 16 recently and promptly
ran out. If we can make it 32 that would be great.

On Oct 30, 2016 2:44 AM, "Alden Hart" <notifications@github.com> wrote:

@lutorm <https://github.com/lutorm> , I'm working on getting all this out
for test. I have your code installed and have been running it. I will post
as a dev branch, probably something like dev-109-tool-offset - hopefully
today if I do not get side tracked.

One question - I set the tool table size to 16. Is this adequate for your
uses, or do you need it to be larger? I can expand to 24 or 32, but bigger
than this is problematic.

—
You are receiving this because you were mentioned.
Reply to this email directly, view it on GitHub
<https://github.com/synthetos/g2/pull/109#issuecomment-257148936>, or mute
the thread

**Comment by aldenhart** [comment]
@lutorm : I posted the code to the dev-109-tool-offset feature branch. Consider this alpha. The tool table has been expanded to 32 tools. I have not had a chance to develop a complete functional/regression test for this yet, but I do have enough tests running to se that it should be reproducing your original code. Documentation can be found here: https://github.com/synthetos/g2/wiki/Tool-Selection-and-Offsets

I am interested n collaborating on the M6 manual tool change. I have some ideas, but would like to find out where you are. Perhaps a hangout is in order.

Please let me know what you find.

**Comment by lutorm** [comment]
Ok, I'm trying it out. I think I managed to port my settings to the new branch. However, I'm having a problem with the communication. I don't get any responses to e.g. "$x" to see settings like I did before. I remember reading something about you redesigning the comms, so is there something I need to do with chilipeppr or the serial port json server for it to echo responses? Or is there a setting for this?

**Comment by aldenhart** [comment]
The EJ command has changed - see this link:
https://github.com/synthetos/g2/wiki/Configuring-0.99-System-Groups#ej-enable-json-mode
You might need to try it in auto mode, ej=2

I'll see what I can find out about connecting to SPJS

**Comment by lutorm** [comment]
Ah, ok. With ej=2 it responds to config requests like it used to. However, then the chilipeppr feedback to moves doesn't work, because if I do a "g0 x5", it echoes the position readouts in text mode... No big deal, I can proceed without that, but I hope the old behavior can be restored somehow.

**Comment by chilipeppr** [comment]
Remember that SPJS TinyG buffer shoves an {"ej":""} after any text command it sees to force TinyG back into JSON mode because without JSON mode active all the time ChiliPeppr breaks. It used to be on v8 that the response would come back in text mode fast enough that the {"ej":""} would only affect commands afterwards. Seems new build processes {"ej":""} faster and thus we don't get that anymore.

It was sorta lame I had to coerce TinyG back into JSON mode all the time though in the TinyG buffer in SPJS. I never liked that. So i like that ej has more options now.

**Comment by lutorm** [comment]
I'm running into more problems -- jogging hangs the connection and the SPJS. It doesn't even respond to kill -9.  I guess I should create these as chilipeppr issues?

**Comment by chilipeppr** [comment]
Before you consider it being SPJS, jogging makes ample use of feed hold and I have found historically that this crashes TinyG a lot.

**Comment by lutorm** [comment]
@chilipeppr Maybe, but it still shouldn't hang like that. I have to reboot to get rid of it.

**Comment by lutorm** [comment]
I tried the things I use for my normal workflow:
 * Insert a pre-measured tool and G43 to set its offset.
 * That tool offsets affect all coordinate systems.
 * Insert an un-measured tool, jog to a known position using the Z-axis setter, and G10 L10 to that position.
 * G49 cancels offset.

Those all seem to behave the way they should (or at least the way they did before, I'm unclear on whether a G10 L10 of the current tool should be reflected immediately or only after a G43. The code I wrote wasn't immediate, but then I read something that made me think maybe it should be. In any case, that's not an immediate concern.)

So this looks good to me. Now if I can only figure out this jogging hang. It doesn't happen when you jog with the UI buttons, which do not send a feed hold, so this appears to support @chilipeppr's theory that it's the feedhold. When I used the keyboard, I get the hang on the second move, with an exception message. I'll make a separate issue for this.

**Comment by chilipeppr** [comment]
I have seen tons of crashing on feed hold in G2 core. When using v8 I never see crashing on feed hold.

**Comment by aldenhart** [comment]
I have been running your files on g2 and torture testing feedhold with no issues. Let's compare notes offline.

**Comment by aldenhart** [comment]
These changes have been merged into edge. It now has a 32 item tool table. Please let me know if you have any questions. I'm closing this PR.

Docs are here: https://github.com/synthetos/g2/wiki/Tool-Selection-and-Offsets

**Pull request closed without merging**